src/System.Net.Security/src/System/Net/_SSPIWrapper.cs

```diff
@@ -0,0 +1,873 @@
+            // const int SECPKG_NEGOTIATION_DIRECT          = 3;
+            // const int SECPKG_NEGOTIATION_TRY_MULTICRED   = 4;
+
+            if (negotiationState == SECPKG_NEGOTIATION_COMPLETE || negotiationState == SECPKG_NEGOTIATION_OPTIMISTIC)
+            {
+                IntPtr unmanagedString = Marshal.ReadIntPtr(packageInfo, SecurityPackageInfo.NameOffest);
+                string name = null;
+                if (unmanagedString != IntPtr.Zero)
+                {
+                    name = Marshal.PtrToStringUni(unmanagedString);
+                }
+                GlobalLog.Print("NegotiationInfoClass::.ctor() packageInfo:" + packageInfo.ToString("x8") + " negotiationState:" + negotiationState.ToString("x8") + " name:" + Logging.ObjectToString(name));
+
+                // an optimization for future string comparisons
+                if (string.Compare(name, Kerberos, StringComparison.OrdinalIgnoreCase) == 0)
+                {
+                    AuthenticationPackage = Kerberos;
+                }
+                else if (string.Compare(name, NTLM, StringComparison.OrdinalIgnoreCase) == 0)
+                {
+                    AuthenticationPackage = NTLM;
+                }
+                else if (string.Compare(name, WDigest, StringComparison.OrdinalIgnoreCase) == 0)
+                {
+                    AuthenticationPackage = WDigest;
+                }
+                else
+                {
+                    AuthenticationPackage = name;
+                }
+            }
+        }
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct SecurityPackageInfo
+    {
+        // see SecPkgInfoW in <sspi.h>
+        internal int Capabilities;
+        internal short Version;
+        internal short RPCID;
+        internal int MaxToken;
+        internal IntPtr Name;
+        internal IntPtr Comment;
+
+        internal static readonly int Size = Marshal.SizeOf<SecurityPackageInfo>();
+        internal static readonly int NameOffest = (int)Marshal.OffsetOf<SecurityPackageInfo>("Name");
+    }
+
+    internal class SecurityPackageInfoClass
+    {
+        internal int Capabilities = 0;
+        internal short Version = 0;
+        internal short RPCID = 0;
+        internal int MaxToken = 0;
+        internal string Name = null;
+        internal string Comment = null;
+
+        /*
+         *  This is to support SSL under semi trusted environment.
+         *  Note that it is only for SSL with no client cert
+         */
+        internal SecurityPackageInfoClass(SafeHandle safeHandle, int index)
+        {
+            if (safeHandle.IsInvalid)
+            {
+                GlobalLog.Print("SecurityPackageInfoClass::.ctor() the pointer is invalid: " + (safeHandle.DangerousGetHandle()).ToString("x"));
+                return;
+            }
+            IntPtr unmanagedAddress = IntPtrHelper.Add(safeHandle.DangerousGetHandle(), SecurityPackageInfo.Size * index);
```

Comment:
Verified callers: the safe handle is created by the caller and kept alive, on the stack for the duration of this ctor. I've added a comment.